mistralrs-core/src/vision_models/llava/llava15.rs

```diff
@@ -0,0 +1,333 @@
+        let mm_projector = MMProjector::new(&vb, config)?;
+        let clip_vision_tower = ClipVisionTower::new(
+            vb.pp("vision_tower.vision_model"),
+            config.vision_feature_layer,
+            &config.vision_feature_select_strategy,
+            &clip_config,
+        )?;
+
+        let llm: Box<dyn LLaVALLM> = match config.text_config.model_type.as_str() {
+            "llama" => {
+                let llama_config = config.to_llama_config();
+                let llama = Llama::new(
+                    &llama_config,
+                    vb.pp("language_model"),
+                    is_gptx,
+                    normal_loading_metadata,
+                )?;
+                Box::new(llama)
+            }
+            "mistral" => {
+                let mistral_config = config.to_mistral_config();
+                let mistral = Mistral::new(
+                    &mistral_config,
+                    vb.pp("language_model"),
+                    is_gptx,
+                    normal_loading_metadata,
+                )?;
+                Box::new(mistral)
+            }
+            _ => {
+                bail!("Unsupported model type: {}", config.text_config.model_type);
+            }
+        };
+        Ok(Self {
+            clip_vision_tower,
+            mm_projector,
+            llm,
+            config: config.clone(),
+            device,
+            dtype,
+        })
+    }
+
+    pub fn encode_images(&self, x: &Tensor) -> Result<Tensor> {
+        let mut image_features = self.clip_vision_tower.forward(x)?;
+        image_features = self.mm_projector.forward(&image_features)?;
+        Ok(image_features)
+    }
+
+    pub fn prepare_inputs_labels_for_multimodal(
+        &self,
+        input_ids: &Tensor, //[1,seq_len]
+        images: &Tensor,    //[sum of samples of all images,channel,width,height]
+        num_image_tokens: usize,
+    ) -> Result<Tensor> {
+        let image_indexes = input_ids
+            .squeeze(0)?
+            .lt(0i64)?
+            .nonzero()?
+            .squeeze(1)?
+            .to_vec1::<u32>()?;
+        let image_ids = image_indexes
+            .iter()
+            .map(|x| Ok(-(input_ids.i((0, *x as usize))?.to_scalar::<i64>()?)))
+            .collect::<Result<Vec<i64>>>()?;
+        let mut result = input_ids.clamp(0i64, i64::MAX)?.to_dtype(DType::U32)?;
+        result = self.llm.embed(&result)?; //[seq_len,hidden_size]
+        let image_features = self.encode_images(&images.to_dtype(self.dtype)?)?; //[num of images,patch_size*patch_size,hidden_size]
+        let num_of_images = image_features.shape().dims()[0];
+        let mut image_features_vec = Vec::new();
+        for i in 0..num_of_images {
+            image_features_vec.push(image_features.get(i)?.unsqueeze(0)?);
+        }
+        for (i, image_index) in image_indexes.iter().enumerate() {
+            let image_id = image_ids[i];
+            result = result.slice_assign(
```

Comment:
What were your thoughts on our slice assign changes on the Candle fork 😃? I have a Candle PR open.